cmd/fyne/internal/commands/translate.go

```diff
@@ -0,0 +1,374 @@
+			&cli.BoolFlag{
+				Name:    "verbose",
+				Aliases: []string{"v"},
+				Usage:   "Show files that are being scanned etc.",
+			},
+			&cli.BoolFlag{
+				Name:    "update",
+				Aliases: []string{"u"},
+				Usage:   "Update existing translations (use with care).",
+			},
+			&cli.BoolFlag{
+				Name:    "dry-run",
+				Aliases: []string{"n"},
+				Usage:   "Scan without storing the results.",
+			},
+			&cli.StringFlag{
+				Name:    "sourceDir",
+				Aliases: []string{"d"},
+				Usage:   "Directory to scan recursively for go files.",
+				Value:   ".",
+			},
+			&cli.StringFlag{
+				Name:    "translationsFile",
+				Aliases: []string{"f"},
+				Usage:   "File to read from and write translations to.",
+				Value:   "translations/en.json",
+			},
+		},
+		Action: func(ctx *cli.Context) error {
+			sourceDir := ctx.String("sourceDir")
+			translationsFile := ctx.String("translationsFile")
+			files := ctx.Args().Slice()
+			opts := translateOpts{
+				DryRun:  ctx.Bool("dry-run"),
+				Update:  ctx.Bool("update"),
+				Verbose: ctx.Bool("verbose"),
+			}
+
+			// without any argument find all .go files in the source directory
+			if len(files) == 0 {
+				sources, err := findFilesExt(sourceDir, ".go")
+				if err != nil {
+					return err
+				}
+				files = sources
+			}
+
+			// update the translation file by scanning the given source files
+			return updateTranslationsFile(&opts, translationsFile, files)
+		},
+	}
+}
+
+// Recursively walk the given directory and return all files with the matching extension
+func findFilesExt(dir, ext string) ([]string, error) {
+	files := []string{}
+	err := filepath.Walk(dir, func(path string, fi fs.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if filepath.Ext(path) != ext || fi.IsDir() || !fi.Mode().IsRegular() {
+			return nil
+		}
+
+		files = append(files, path)
+
+		return nil
+	})
+	return files, err
+}
+
+type translateOpts struct {
+	DryRun  bool
+	Update  bool
+	Verbose bool
+}
+
+// Create or add to translations file by scanning the given files for translation calls
+func updateTranslationsFile(opts *translateOpts, file string, files []string) error {
+	translations := make(map[string]interface{})
+
+	// try to get current translations first
+	f, err := os.Open(file)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return err
+	}
+	defer f.Close()
+
+	// only try to parse translations that exist
+	if f != nil {
+		dec := json.NewDecoder(f)
+		if err := dec.Decode(&translations); err != nil {
+			return err
+		}
+	}
+
+	//if ctx.Bool("verbose") {
+	if opts.Verbose {
+		fmt.Fprintf(os.Stderr, "scanning files: %v\n", files)
+	}
+
+	// update translations hash
+	if err := updateTranslationsHash(opts, translations, files); err != nil {
+		return err
+	}
+
+	// serialize in readable format for humas to change
+	b, err := json.MarshalIndent(translations, "", "\t")
+	if err != nil {
+		return err
+	}
+
+	// avoid writing an empty file
+	if len(translations) == 0 {
```

Comment:
Should this happen before json encoding?

Reply:
Yes, good catch.